web/src/composebox_typeahead.ts

```diff
@@ -1072,167 +1073,184 @@ export function content_typeahead_selected(
             message: "",
         };
     }
-    switch (item.type) {
-        case "emoji":
-            // leading and trailing spaces are required for emoji,
-            // except if it begins a message or a new line.
-            if (
-                beginning.lastIndexOf(":") === 0 ||
-                beginning.charAt(beginning.lastIndexOf(":") - 1) === " " ||
-                beginning.charAt(beginning.lastIndexOf(":") - 1) === "\n"
-            ) {
-                beginning = beginning.slice(0, -token.length - 1) + ":" + item.emoji_name + ": ";
-            } else {
-                beginning = beginning.slice(0, -token.length - 1) + " :" + item.emoji_name + ": ";
-            }
-            break;
-        case "user_group":
-        case "user":
-        case "broadcast": {
-            const is_silent = item.is_silent;
-            beginning = beginning.slice(0, -token.length - 1);
-            if (beginning.endsWith("@_*")) {
-                beginning = beginning.slice(0, -3);
-            } else if (beginning.endsWith("@*") || beginning.endsWith("@_")) {
-                beginning = beginning.slice(0, -2);
-            } else if (beginning.endsWith("@")) {
-                beginning = beginning.slice(0, -1);
-            }
-            if (item.type === "user_group") {
-                let user_group_mention_text = is_silent ? "@_*" : "@*";
-                user_group_mention_text += item.name + "* ";
-                beginning += user_group_mention_text;
-                // We could theoretically warn folks if they are
-                // mentioning a user group that literally has zero
-                // members where we are posting to, but we don't have
-                // that functionality yet, and we haven't gotten much
-                // feedback on this being an actual pitfall.
-            } else {
-                const user_id = item.type === "broadcast" ? undefined : item.user.user_id;
-                let mention_text = people.get_mention_syntax(
-                    item.user.full_name,
-                    user_id,
-                    is_silent,
-                );
-                if (!is_silent && item.type !== "broadcast") {
-                    compose_validate.warn_if_mentioning_unsubscribed_user(item, $textbox);
-                    mention_text = compose_validate.convert_mentions_to_silent_in_direct_messages(
-                        mention_text,
+    // We only want to consider escape key for the stream+topic typeahead completion case.
+    if (event?.key !== "Escape" || item.type === "topic_list") {
```

Comment:
Could this use the early return pattern, i.e. a conditional of the form `event?.key === "Escape" && ...` that just handles that case, and then let it fall through to the switch the non-escape case?

It's a shame to indent so much code deeper.